### PR TITLE
Re-apply "ModuleInterface: lock .swiftinterface while generating module cache"

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: lib/Frontend/ModuleInterfaceBuilder.cpp [hunk 1]
```diff
@@ -384,3 +385,70 @@ bool ModuleInterfaceBuilder::buildSwiftModule(
...
+    if (RemarkRebuild) {
+      RemarkRebuild();
+    }
+    return buildSwiftModuleInternal(OutPath, ShouldSerializeDeps, ModuleBuffer);
+  }
+  case llvm::LockFileManager::LFS_Shared: {
+    // Someone else is responsible for building the module. Wait for them to
+    // finish.
+    switch (Locked.waitForUnlock()) {
+    case llvm::LockFileManager::Res_Success: {
```
Comment (CodaFi) [inline lib/Frontend/ModuleInterfaceBuilder.cpp]: This is still going to cause stampedes.  A lot of the interfaces take longer than the default timeout interval of 40 seconds to generate.

**Comment by nkcsgexi** [comment]
@swift-ci please test

File: lib/Frontend/ModuleInterfaceBuilder.cpp (hunk 1 above)
Comment (nkcsgexi) [reply lib/Frontend/ModuleInterfaceBuilder.cpp]: We could change llvm side to take a parameter for `waitForUnlock()` to indicate the maximum waiting time. It doesn't fundamentally invalidate this approach.

**Pull request merged**